kernel/src/fs/utils/flock.rs

```diff
@@ -96,49 +96,65 @@ impl Debug for FlockItem {
 pub struct FlockList {
-    inner: Mutex<VecDeque<FlockItem>>,
+    inner: RwMutex<VecDeque<FlockItem>>,
 }
 
 impl FlockList {
     /// Creates a new FlockList.
     pub fn new() -> Self {
         Self {
-            inner: Mutex::new(VecDeque::new()),
+            inner: RwMutex::new(VecDeque::new()),
         }
     }
 
     /// Attempts to set a lock on the file.
+    ///
+    /// If no conflicting locks exist, the lock is set and the function returns Ok(()).
+    /// If a conflicting lock exists:
+    /// - If waker is not None, it is added to the conflicting lock's waitqueue, and the function returns EAGAIN.
+    /// - If waker is None, the function returns EAGAIN.
+    fn try_set_lock(&self, req_lock: &FlockItem, waker: Option<Arc<Waker>>) -> Result<()> {
+        let mut list = self.inner.write();
+        if let Some(conflict_lock) = list.iter().find(|l| req_lock.conflict_with(l)) {
+            if let Some(waker) = waker {
+                conflict_lock.waitqueue.enqueue(waker);
+            }
+            return_errno_with_message!(Errno::EAGAIN, "the file is locked");
+        } else {
+            match list.iter().position(|l| req_lock.same_owner_with(l)) {
+                Some(idx) => {
+                    list[idx] = req_lock.clone();
+                }
+                None => {
+                    list.push_front(req_lock.clone());
+                }
+            }
+            Ok(())
+        }
+    }
+
+    /// Set a lock on the file.
+    ///
     /// If no conflicting locks exist, the lock is set and the function returns Ok(()).
     /// If is_nonblocking is true and a conflicting lock exists, the function returns EAGAIN.
-    /// Otherwise, the function waits until the lock can be acquired.
-    pub fn set_lock(&self, mut req_lock: FlockItem, is_nonblocking: bool) -> Result<()> {
+    /// Otherwise, the function waits until the lock can be acquired or until it is interrupted by a signal.
+    pub fn set_lock(&self, req_lock: FlockItem, is_nonblocking: bool) -> Result<()> {
         debug!(
             "set_lock with Flock: {:?}, is_nonblocking: {}",
             req_lock, is_nonblocking
         );
-        loop {
-            let (waiter, waker);
-            {
-                let mut list = self.inner.lock();
-                if let Some(existing_lock) = list.iter().find(|l| req_lock.conflict_with(l)) {
-                    if is_nonblocking {
-                        return_errno_with_message!(Errno::EAGAIN, "the file is locked");
-                    }
-                    (waiter, waker) = Waiter::new_pair();
-                    existing_lock.waitqueue.enqueue(waker);
+        if is_nonblocking {
+            self.try_set_lock(&req_lock, None)
+        } else {
+            let (waiter, waker) = Waiter::new_pair();
+            waiter.pause_until(|| {
+                let result = self.try_set_lock(&req_lock, Some(waker.clone()));
+                if result.is_err() {
```

Comment:
```suggestion
               result.is_err_and(|err| err.errno() == Errno::EAGAIN)
```

`try_set_lock` does not currently return an error code other than `EAGAIN`, this is just in case future changes make it possible. This should be cheap and harmless enough.

Reply:
Using `result.is_err_and(|err| err.errno() == Errno::EAGAIN)` to break the wait is correct. But if `try_set_lock()` can return other error codes, like `ENOLCK` when the kernel runs out of memory, how should we handle these errors? `pause_until()` seems to return `Ok(())`.

Reply:
Maybe `wait_events()` should be used in such case?

Reply:
We're likely to add deadlock detection for `range_lock` (https://github.com/asterinas/asterinas/issues/1467), so `try_set_lock()` will return `EDEADLK`. We should handle this case.

Reply:
> Using `result.is_err_and(|err| err.errno() == Errno::EAGAIN)` to break the wait is correct. But if `try_set_lock()` can return other error codes, like `ENOLCK` when the kernel runs out of memory, how should we handle these errors? `pause_until()` seems to return `Ok(())`.

We need to return `Some(result)`, and then call [`Result::flatten`](https://doc.rust-lang.org/std/result/enum.Result.html#method.flatten) after `pause_until`.

> Maybe `wait_events()` should be used in such case?

It is semantically similar. However, there are no pollable instances, so we cannot use it.